code/datums/emotes.dm

```diff
@@ -0,0 +1,165 @@
+#define EMOTE_VISIBLE 1
+#define EMOTE_AUDIBLE 2
+
+/datum/emote
+	var/key = "" //What calls the emote
+	var/key_third_person = "" //This will also call the emote
+	var/message = "" //Message displayed when emote is used
+	var/message_param = "" //Message to display if a param was given
+	var/emote_type = EMOTE_VISIBLE //Whether the emote is visible or audible
+	var/restraint_check = FALSE //Checks if the mob is restrained before performing the emote
+	var/muzzle_ignore = FALSE //Will only work if the emote is EMOTE_AUDIBLE
+	var/muzzled_noise = "" // For snowflaking the muzzled noise
+	var/list/mob_type_blacklist_typecache //Types that are NOT allowed to use that emote
+	var/list/mob_type_ignore_stat_typecache
+	var/punct = "."
+	var/stat_allowed = CONSCIOUS
+	var/cooldown = 0
+	var/sound
+	var/sound_volume = 50
+	var/sound_vary = 0
+	var/sound_frequency = 0
+	var/static/list/emote_lists = list(
+		"alien"=init_subtypes(/datum/emote/alien),
+		"brain"=init_subtypes(/datum/emote/brain),
+		"carbon"=init_subtypes(/datum/emote/carbon),
+		"human"=init_subtypes(/datum/emote/human),
+		"living"=init_subtypes(/datum/emote/living),
+		"mob"=init_subtypes(/datum/emote/mob),
+		"nymph"=init_subtypes(/datum/emote/nymph),
+		"robot"=init_subtypes(/datum/emote/robot),
+		"silicon"=init_subtypes(/datum/emote/silicon),
+		"simple_animal"=init_subtypes(/datum/emote/simple_animal),
+		"slime"=init_subtypes(/datum/emote/slime),
+		"synth"=init_subtypes(/datum/emote/synth),
+	)
+
+/datum/emote/New()
+	mob_type_blacklist_typecache = typecacheof(mob_type_blacklist_typecache)
+	mob_type_ignore_stat_typecache = typecacheof(mob_type_ignore_stat_typecache)
+
+/datum/emote/proc/run_emote(mob/user, params, type_override)
+	if(!can_run_emote(user))
+		return FALSE
+
+	var/msg = create_emote_message(user, params)
+	if(!msg)
+		return TRUE
+	show_to_ghosts(user, msg)
+
+	var/ET = type_override ? type_override : emote_type
+
+	if(ET == EMOTE_AUDIBLE)
+		user.audible_message(msg)
+		if(can_play_sound(user))
+			play_sound(user)
+	else
+		user.visible_message(msg)
+		play_sound(user)
+	log_emote(msg, user)
+	emote_cooldown(user)
+	return TRUE
+
+// Override this to use custom message behavior, or do something before the emote is printed to the screen
+/datum/emote/proc/create_emote_message(mob/user, params)
+	var/msg = select_message_type(user)
+	if(params)
+		msg = select_param(user, params, msg)
+
+	msg = replace_pronoun(user, msg)
+
+	if(!msg)
+		return
+
+	msg = "<b>[user]</b> " + msg + punct
+	return msg
+
+/datum/emote/proc/show_to_ghosts(mob/user, msg)
+	if(findtext(msg, " snores."))
+		return
+
+	for(var/mob/M in player_list)
+		if(!M.client || isnewplayer(M))
+			continue
+		var/T = get_turf(user)
+		if(M.stat == DEAD && M.client && M.get_preference(CHAT_GHOSTSIGHT) && !(M in viewers(T, null)))
+			M.show_message(msg)
+
+/datum/emote/proc/emote_cooldown(mob/user)
+	if(!cooldown || user.emote_on_cd)
+		return
+	user.emote_on_cd = TRUE
```

Comment:
Why does this use persistent variables instead of a cooldown system like movement or clicking? Just a global `if(user.emote_cd > world.time) return` and then `adjustEmoteCooldown(+50 ticks)`